src/crypto/CryptoInterface.c

```diff
@@ -81,57 +81,90 @@ void BSL_Crypto_SetRngGenerator(BSL_Crypto_RandBytesFn rand_gen_fn)
     rand_bytes_generator = rand_gen_fn;
 }
 
-int BSL_Crypto_UnwrapKey(BSL_Data_t *unwrapped_key_output, BSL_Data_t wrapped_key_plaintext, const char *key_id,
-                         size_t aes_variant)
+int BSL_Crypto_ClearKeyHandle(void *keyhandle)
 {
+    CHK_ARG_NONNULL(keyhandle);
+
+    BSLB_CryptoKey_t *key = (BSLB_CryptoKey_t *)keyhandle;
+    CHK_POSTCONDITION(BSL_SUCCESS == BSLB_CryptoKey_Deinit(key));
+    BSL_FREE(key);
+
+    return BSL_SUCCESS;
+}
+
+int BSL_Crypto_UnwrapKey(const void *kek_handle, size_t aes_variant, BSL_Data_t *wrapped_key, const void **cek_handle)
+{
+    BSLB_CryptoKey_t *kek = (BSLB_CryptoKey_t *)kek_handle;
+
+    BSLB_CryptoKey_t *cek = BSL_MALLOC(sizeof(BSLB_CryptoKey_t));
+
     const EVP_CIPHER *cipher = (aes_variant == BSL_CRYPTO_AES_128) ? EVP_aes_128_wrap() : EVP_aes_256_wrap();
     EVP_CIPHER_CTX   *ctx    = EVP_CIPHER_CTX_new();
-    ASSERT_ARG_NONNULL(ctx);
+    if (ctx == NULL)
+    {
+        BSL_FREE(cek);
+        return BSL_ERR_SECURITY_CONTEXT_CRYPTO_FAILED;
+    }
 
-    // Give the actual key extra margin on each side.
-    uint8_t  keybuf[128];
-    uint8_t *key = &keybuf[16];
-    memset(keybuf, 0, sizeof(keybuf));
+    BSL_Data_InitBuffer(&cek->raw, kek->raw.len);
 
-    size_t keylen = 0;
-    if (BSL_SUCCESS != BSLB_Crypto_GetRegistryKey(key_id, (const uint8_t **)&key, &keylen))
+    int dec_result = EVP_DecryptInit_ex(ctx, cipher, NULL, kek->raw.ptr, NULL);
+    if (dec_result != 1)
     {
+        BSL_Data_Deinit(&cek->raw);
+        BSL_FREE(cek);
         return BSL_ERR_SECURITY_CONTEXT_CRYPTO_FAILED;
     }
-    ASSERT_POSTCONDITION(keylen > 0);
-
-    int dec_result = EVP_DecryptInit_ex(ctx, cipher, NULL, key, NULL);
-    ASSERT_POSTCONDITION(dec_result == 1);
     EVP_CIPHER_CTX_set_padding(ctx, 0);
 
-    int unwrapped_key_len = 16;
-    int decrypt_res = EVP_DecryptUpdate(ctx, unwrapped_key_output->ptr, &unwrapped_key_len, wrapped_key_plaintext.ptr,
-                                        (int)wrapped_key_plaintext.len);
+    int decrypt_res = EVP_DecryptUpdate(ctx, cek->raw.ptr, (int *)&cek->raw.len, wrapped_key->ptr, wrapped_key->len);
     if (decrypt_res != 1)
     {
         BSL_LOG_ERR("EVP_DecryptUpdate: %s", ERR_error_string(ERR_get_error(), NULL));
+        BSL_Data_Deinit(&cek->raw);
+        BSL_FREE(cek);
         EVP_CIPHER_CTX_free(ctx);
-        return -1;
+        return BSL_ERR_SECURITY_CONTEXT_CRYPTO_FAILED;
     }
 
-    unwrapped_key_output->len = (size_t)unwrapped_key_len;
-
     int final_len = 0;
-    int res       = EVP_DecryptFinal_ex(ctx, &unwrapped_key_output->ptr[unwrapped_key_output->len], &final_len);
+    int res       = EVP_DecryptFinal_ex(ctx, &wrapped_key->ptr[wrapped_key->len], &final_len);
```

Comment:
I don't think this is right; it may just not be erroring because this mechanism doesn't use padding. The second argument is supposed to be a buffer (of some minimum size) and this will give a pointer to a zero-length. If possible it may just be better to give the NULL pointer here.